### PR TITLE
Post Editor: reduxify the publish button

### PR DESCRIPTION
**Pull request by jsnajdr**
This PR introduces a new selector named `getEditorPublishButtonStatus` and uses it to reduxify the publish buttons: the one in editor sidebar:

screen shot 2018-05-22 at 14 30 18

and also the one in the confirmation sidebar, if enabled:

screen shot 2018-05-22 at 14 43 33

The patch is divided into several incremental commits: read the commit messages for details.

**How to test:**
The label of the publish button should reflect the action that is going to happen if you press it: for drafts that are going to be published, it's "Publish". For posts with future publish date it's "Schedule". For already published posts, it's "Update". For users that can't publish and need an approval, it's "Request Review". Verify that all scenarios work.

The button is also enabled only when the post is in a valid state: it needs to have content and a password-protected post must have a valid password. Verify that the button is enabled and disabled as it should be.

When a publish action is in progress, the button should be marked as busy.

### CODE REVIEW

**Comment by jsnajdr** [comment]
All the CircleCI checks are failing miserably, but that's due to issues with NPM registry and CircleCI and whatever else is broken today. I'll restart them in a few hours.

**Comment by alisterscott** [comment]
All the CircleCI tests now passing ✅

**Review by blowery — APPROVED** [review]
LGTM, tested out fine. Would be nice to add one test to cover the `null` return from the new selector, but not essential.

File: client/state/ui/editor/test/selectors.js [hunk 1]
```diff
@@ -325,4 +327,97 @@ describe( 'selectors', () => {
...
+
+		test( 'should return "publish" if the post is a draft', () => {
+			const state = editorState( { status: 'draft' } );
+			expect( getEditorPublishButtonStatus( state ) ).to.equal( 'publish' );
+		} );
+
+		test( 'should return "requestReview" if the post is a draft and user can\'t publish', () => {
+			const state = editorState( { status: 'draft' }, null, false );
+			expect( getEditorPublishButtonStatus( state ) ).to.equal( 'requestReview' );
+		} );
```
[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/state/ui/editor/test/selectors.js]: Sure, I pushed a commit with tests for the `null` case. And also rebased the PR onto the latest master and resolved a few conflicts (all of them in the `import` blocks).

**Comment by damiannep** [comment]
There may be an issue with this. When a user creates a new post, they get the following view. 

nopublish

It changes with the first save as the user enters content. Is this the expected behavior? Posted [here](https://github.com/Automattic/wp-calypso/issues/25032).

**Pull request merged**